internal/controller/siteconfig_builder.go

```diff
@@ -37,16 +42,113 @@ func NewSiteConfigBuilder(pLog logr.Logger) *SiteConfigBuilder {
 	return &scBuilder
 }
 
-// getConfigMap retrieves the configmap from cluster for a given TemplateRef
-func getConfigMap(ctx context.Context, c client.Client, templateRef v1alpha1.TemplateRef) (*corev1.ConfigMap, error) {
+func (scbuilder *SiteConfigBuilder) ProcessTemplates(ctx context.Context, c client.Client, siteConfig v1alpha1.SiteConfig) ([]interface{}, error) {
+	scbuilder.Log.Info(fmt.Sprintf("Processing cluster-level templates for SiteConfig %s", siteConfig.Name))
+
+	// Render cluster-level templates
+	siteManifests, err := scbuilder.renderTemplates(ctx, c, siteConfig, siteConfig.Spec.TemplateRefs, NonNodeIndex, siteConfig.Spec.SuppressedManifests)
+	if err != nil {
+		scbuilder.Log.Info(fmt.Sprintf("encountered error while processing cluster-level templates for SiteConfig %s, err: %s", siteConfig.Name, err.Error()))
+		return siteManifests, err
+	}
+	scbuilder.Log.Info(fmt.Sprintf("Processed cluster-level templates for SiteConfig %s", siteConfig.Name))
+
+	// Process node-level templates
+	numNodes := len(siteConfig.Spec.Nodes)
+	for nodeId, node := range siteConfig.Spec.Nodes {
+		scbuilder.Log.Info(fmt.Sprintf("Processing node-level templates for SiteConfig %s [node: %d of %d]", siteConfig.Name, nodeId+1, numNodes))
+
+		// Render node-level templates
+		nodeManifests, err := scbuilder.renderTemplates(ctx, c, siteConfig, node.TemplateRefs, nodeId, siteConfig.Spec.Nodes[nodeId].SuppressedManifests)
+		if err != nil {
+			scbuilder.Log.Info(fmt.Sprintf("encountered error while processing node-level templates for SiteConfig %s [%d of %d], err: %s", siteConfig.Name, nodeId+1, numNodes, err.Error()))
+			return siteManifests, err
+		}
+		scbuilder.Log.Info(fmt.Sprintf("Processed node-level templates for SiteConfig %s [node: %d of %d]", siteConfig.Name, nodeId+1, numNodes))
+
+		for _, nodeCR := range nodeManifests {
+			if nodeCR != nil {
+				siteManifests = append(siteManifests, nodeCR)
+			}
+		}
+	}
+
+	return siteManifests, nil
+}
+
+func (scbuilder *SiteConfigBuilder) renderTemplates(ctx context.Context, c client.Client, siteConfig v1alpha1.SiteConfig,
+	templateRefs []v1alpha1.TemplateRef, nodeId int, suppressedManifests []string) ([]interface{}, error) {
```

Comment:
Passing the index of the node around here seems very strange to me.

It looks like it's only used to index into the nodes list in the siteconfig.
Maybe just pass a reference to the node itself rather than doing all this indexing?
Then also if the node is nil you know you're dealing with a cluster and you can get rid of the strange `NonNodeIndex` const which feels like it's going to get you a out of range index error bug at some point.

Reply:
Either way it looks like there's no reason to pass the `suppressedManifests` separately if you're already referencing the node. That data's in the same place, right?

Reply:
refactored code,`templateRefs` and `suppressedManifests` are now determined within the function.